From 4962b936c9e14ec58439b24017a8503d9e0d1416 Mon Sep 17 00:00:00 2001
From: Lei Li <lei.l.li@intel.com>
Date: Thu, 31 Jan 2013 13:01:14 +0800
Subject: [PORT FROM MAIN] fix the menu display issue under recovery mode

BZ: 58239

The root cause is that the compiled code is not cached successfully,
although it's generated successfully.

With this patch, as long as the index of the cached entry is within
the capacity, the return value should be 0. Otherwise it's the real error.

Change-Id: If4838ff06aa1a64e6d535be19bf607064edf31db
Orig-Change-Id: Ie557da79e668c6779b792655ca009c28e4a3b125
Signed-off-by: Lei Li <lei.l.li@intel.com>
Reviewed-on: http://android.intel.com:8080/90093
Reviewed-by: Castelain, DavidX <davidx.castelain@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Reviewed-by: Chen, HuaqiangX <huaqiangx.chen@intel.com>
Reviewed-by: Xu, Randy <randy.xu@intel.com>
Tested-by: Xu, Randy <randy.xu@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 libpixelflinger/codeflinger/CodeCache.cpp |    9 ++++++++-
 1 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/libpixelflinger/codeflinger/CodeCache.cpp b/libpixelflinger/codeflinger/CodeCache.cpp
index 58fde7e..91452e0 100644
--- a/libpixelflinger/codeflinger/CodeCache.cpp
+++ b/libpixelflinger/codeflinger/CodeCache.cpp
@@ -209,9 +209,16 @@ int CodeCache::cache(  const AssemblyKeyBase& keyBase,
                  strerror(errno));
 #endif
     }
+    size_t count = mCacheData.size();
 
     pthread_mutex_unlock(&mLock);
-    return err;
+    /**** For ARM, return value is always 0 as long as cacheflush succeeds. For non-ARM(e.g. x86), err actually indicates the index
+     of the cache entry. As long as the index is within the capacity, the return value should be 0. Otherwise it's the real error.
+    *****/
+    if (err >= 0 && err < count)
+        return 0;
+    else
+        return err;
 }
 
 // ----------------------------------------------------------------------------
-- 
1.7.4.1

